libusb/os/linux_android_jni.c

```diff
@@ -0,0 +1,1054 @@
+	*descriptors = usbi_reallocf(*descriptors, *descriptors_len);
+	if (!*descriptors)
+		return LIBUSB_ERROR_NO_MEM;
+
+	desc = (struct usbi_interface_descriptor){
+		.bLength = LIBUSB_DT_INTERFACE_SIZE,
+		.bDescriptorType = LIBUSB_DT_INTERFACE,
+		.bInterfaceNumber =
+			(*jni_env)->CallIntMethod(jni_env,
+				interface, jni->UsbInterface_getId),
+		.bAlternateSetting =
+			(*jni_env)->CallIntMethod(jni_env,
+				interface, jni->UsbInterface_getAlternateSetting),
+		.bNumEndpoints = num_epoints,
+		.bInterfaceClass = 
+			(*jni_env)->CallIntMethod(jni_env,
+				interface, jni->UsbInterface_getInterfaceClass),
+		.bInterfaceSubClass = 
+			(*jni_env)->CallIntMethod(jni_env,
+				interface, jni->UsbInterface_getInterfaceSubclass),
+		.bInterfaceProtocol =
+			(*jni_env)->CallIntMethod(jni_env,
+				interface, jni->UsbInterface_getInterfaceProtocol),
+		.iInterface = android_jni_gen_string(jni_env,
+			strings, strings_len,
+			(*jni_env)->CallObjectMethod(jni_env,
+				interface, jni->UsbInterface_getName))
+	};
+	if (desc.iInterface < 0)
+		return desc.iInterface;
+	*(struct usbi_interface_descriptor *)(*descriptors + offset) = desc;
+
+	if (desc.bInterfaceNumber >= *num_ifaces)
+		*num_ifaces = desc.bInterfaceNumber + 1;
+
+	for (idx = 0; idx < num_epoints; idx ++) {
+		/* UsbEndpoint endpoint = interface.getEndpoint(idx); */
+		jobject endpoint =
+			(*jni_env)->CallObjectMethod(jni_env,
+				interface, jni->UsbInterface_getEndpoint, idx);
+
+		r =
+			android_jni_gen_epoint(jni,
+				jni_env, endpoint,
+				descriptors, descriptors_len);
+
+		(*jni_env)->DeleteLocalRef(jni_env, endpoint);
+
+		if (r != LIBUSB_SUCCESS)
+			return r;
+	}
+	return LIBUSB_SUCCESS;
+}
+
+static int android_jni_gen_epoint(struct android_jni_context *jni,
+	JNIEnv *jni_env, jobject endpoint,
+	uint8_t **descriptors, size_t *descriptors_len)
+{
+	size_t offset;
+	struct usbi_endpoint_descriptor desc;
+
+	offset = *descriptors_len;
+	*descriptors_len = offset + LIBUSB_DT_ENDPOINT_SIZE;
+	*descriptors = usbi_reallocf(*descriptors, *descriptors_len);
+	if (!*descriptors)
+		return LIBUSB_ERROR_NO_MEM;
+
+	desc = (struct usbi_endpoint_descriptor){
+		.bLength = LIBUSB_DT_ENDPOINT_SIZE,
+		.bDescriptorType = LIBUSB_DT_ENDPOINT,
+		.bEndpointAddress =
+			(*jni_env)->CallIntMethod(jni_env,
+				endpoint, jni->UsbEndpoint_getAddress),
+		.bmAttributes =
+			(*jni_env)->CallIntMethod(jni_env,
+				endpoint, jni->UsbEndpoint_getAttributes),
+		.wMaxPacketSize = libusb_cpu_to_le16(
+			(*jni_env)->CallIntMethod(jni_env,
+				endpoint, jni->UsbEndpoint_getMaxPacketSize)),
+		.bInterval =
+			(*jni_env)->CallIntMethod(jni_env,
+				endpoint, jni->UsbEndpoint_getInterval)
+	};
+	*(struct usbi_endpoint_descriptor *)(*descriptors + offset) = desc;
+	
+	return LIBUSB_SUCCESS;
+}
+
+static int android_jni_gen_string(JNIEnv *jni_env,
+	char **strings, size_t *strings_len, jstring str)
+{
+	size_t str_len, offset, idx;
+	const uint16_t *str_ptr;
+	struct usbi_string_descriptor *desc;
+
```

Comment:
this is possible due to subsequent calls to `android_jni_gen_string` without checks for reallocation failures that may happen a feww lines below
```suggestion

	if (!*strings)
		return LIBUSB_ERROR_NO_MEM;

```

Reply:
Sorry, your comment starts mid-sentence to me; what is it that is possible?  That value is returned down around line 820 already for me over here, under the same set of conditions.  I'll add a comment that the function is only for #875 .  I was hoping I could help #875 happen and integrate it, but didn't end up being available to do so.

Reply:
I understand now.  Thanks.